dandiapi/api/asset_paths.py

```diff
@@ -0,0 +1,151 @@
+from __future__ import annotations
+
+import os
+
+from django.db import transaction
+from django.db.models import Count, F, QuerySet
+
+from dandiapi.api.models import Asset, AssetPath, AssetPathRelation, Version
+
+####################################################################
+# Dandiset and version deletion will cascade to asset path deletion.
+# Thus, no explicit action is needed for these.
+####################################################################
+
+
+def extract_paths(path: str) -> list[str]:
+    nodepaths: list[str] = path.split('/')
+    for i in range(len(nodepaths))[1:]:
+        nodepaths[i] = os.path.join(nodepaths[i - 1], nodepaths[i])
+
+    return nodepaths
+
+
+def get_root_paths(version: Version) -> QuerySet[AssetPath]:
+    """Return all root paths for a version."""
+    # Use prefetch_related here instead of select_related,
+    # as otherwise the resulting join is very large
+    qs = AssetPath.objects.prefetch_related(
+        'asset',
+        'asset__blob',
+        'asset__embargoed_blob',
+        'asset__zarr',
+    )
+    return qs.filter(version=version).alias(num_parents=Count('parent_links')).filter(num_parents=1)
+
+
+def get_path_children(path: AssetPath) -> QuerySet[AssetPath]:
+    """Get all direct children from an existing path."""
+    qs = AssetPath.objects.select_related(
+        'asset',
+        'asset__blob',
+        'asset__embargoed_blob',
+        'asset__zarr',
+    )
+    path_ids = (
+        AssetPathRelation.objects.filter(parent=path, depth=1)
+        .values_list('child', flat=True)
+        .distinct()
+    )
+    return qs.filter(id__in=path_ids)
```

Comment:
One last nitpick before merging. This queryset is unordered which raises a testing warning:
```
UnorderedObjectListWarning: Pagination may yield inconsistent results with an unordered object_list: <class 'dandiapi.api.models.asset_paths.AssetPath'> QuerySet.
```